### ISSUE TITLE
Applying coverage penalty to Marian NMT

### ISSUE THREAD

**Issue by rajat08** [rajat08/CS591W1_project#1]

**Comment by rajat08** [comment]
Since our task is applying coverage penalty to Marian NMT , we need to find a way to get the decoder attention values from the model.

We are planning to use [HuggingFace's implementation of MarianNMT  ](https://huggingface.co/transformers/model_doc/marian.html)

Though HuggingFace does provide a way to get attention outputs for [BART](https://huggingface.co/transformers/model_doc/bart.html) and [XLM](https://huggingface.co/transformers/model_doc/xlm.html) models they do not provide a out of the box method for Marian NMT itself. But we can use the properties of transformer class and set some flags to get decoder attentions and decoder hidden states. 

Here's what decoder related info we can get-
decoder_hidden_states : Hidden-states of the decoder at the output of each layer plus the initial embedding outputs.
decoder_attentions : Attentions weights of the decoder, after the attention softmax, used to compute the weighted average in the self-attention heads.

Experimenting on this [notebook.](https://colab.research.google.com/drive/1Mb8oHTA_dWx0z8BBOCHK5nvHBzcKCIcX?authuser=1#scrollTo=XhA_z0Q1vdhT)

For applying coverage penalty :
Once we have the weights the decoder attention weights , we can follow the formula mentioned in the paper and in the code here : https://github.com/lvapeab/nmt-keras/blob/interactive_NMT/active_learning_nmt.py
